lib/dalli/protocol/base.rb

```diff
@@ -143,6 +145,14 @@ def quiet?
       end
       alias multi? quiet?
 
+      def pipelined_get_request(keys)
+        req = +''
```

Comment:
I know this line was just copied and moved from pre-existing code, but looking at it we can almost certainly reduce allocations by pre-allocating `req` to a reasonable size.  Something like `keys.size * 24 + keys.map(&:size).sum)` for binary? And something like `keys.size * 18 + keys.map(&:size).sum)` for meta.

Reply:
Will do. But it seems that it should be 17 for meta, at least it was in my tests (non-base64, which will completely break this calculation..).
Does not seem to make a significant difference in speed, though.

I have another idea for a performance optimization: start of by only generating as much bytes as are needed for the first chunk and send it, then pre-calculate upcoming bytes whenever the select blocks (or the socket is writable and no pre-calculated data is available).
This means the server can start generating the response earlier, dalli would use less memory, and time that is "lost" waiting for I/O could be used for calculation.
But it will be a bit harder to do elegantly without a library like async or eventmachine, and I don't know if it is worth the added complexity. Anyway, I would put this in an extra PR once this one is finished.

Reply:
It's very possible I got the number wrong for meta (I did the calculation quickly) but just make sure you're including the separator.

I think the benefit for that optimization is unlikely to be worth the tradeoff, but we can discuss.

Reply:
I think that's a neat idea for an optimization, but yeah it feels like it could be an improvement for a future PR.

The `.map(&:size).sum)` is faster in my little benchmark, but by only by like 10 nanoseconds a key.

Reply:
> The `.map(&:size).sum)` is faster in my little benchmark, but by only by like 10 nanoseconds a key.

I had the (naive) assumption that the cost of allocating an array should be prevented - not knowing that it is outweighed by doing the sum in C instead if in Ruby. I should have remembered Knuth.